### PR TITLE
[java] Modify WhileLoopWithLiteralBoolean to meet the missing case #3455

### LINKED ISSUE #3455: [java] WhileLoopWithLiteralBoolean - false negative with complex expressions
**Affects PMD Version:6.35.0** 

**Rule:WhileLoopWithLiteralBoolean**

**Link:**
https://pmd.github.io/pmd-6.36.0/pmd_rules_java_bestpractices.html#whileloopwithliteralboolean

**Description:**
Hi, when using PMD, I found some false negatives caused by compound but constant expression. I have provided them in the following minimal code sample. Actually, expressions in conditional statements are constant and they should have been detected. Thanks.

**Code Sample demonstrating the issue:**

```java
public void func() {
    do {
      // Loop Body
    } while(false || false); // This is a false negative
//        } while(false | false); // This is a similar false negative.
}
```

**Expected outcome: Two warnings**

PMD should report two violations at line 4 and 5, but doesn't. This is a false-negative.

**Running PMD through:** *[Maven]*

### PR DESCRIPTION
**Pull request by VoidxHoshi**
## Describe the PR
This PR fixes the WhileLoopWithLiteralBoolean rule for 'while( false | false )' case encountered in issue 3455. This parse error has been fixed by changing the grammar in bestpractices.xml file.

## Related issues

- Fixes #3455

## Ready?

- [X] Added unit tests for fixed bug/feature
- [X] Passing all unit tests
- [X] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [X] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 5 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 2 errors and 6 configuration errors.<br />
<a href="https://chunk.io/pmd/84de32cf73d8463c934e520fbe31e1c0/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/WhileLoopWithLiteralBoolean.xml [hunk 1]
```diff
@@ -32,6 +32,34 @@ class Foo {
         ]]></code>
     </test-code>
 
+    <test-code>
+        <description>do while false | false</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/WhileLoopWithLiteralBoolean.xml:36]:
```suggestion
        <description>do while false | false #3455</description>
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/WhileLoopWithLiteralBoolean.xml [hunk 2]
```diff
@@ -32,6 +32,34 @@ class Foo {
...
+    {
+        do {
+        } while (false | false);
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>do while false || false</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/WhileLoopWithLiteralBoolean.xml:50]:
```suggestion
        <description>do while false || false #3455</description>
```

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 3]
```diff
@@ -2171,21 +2171,27 @@ a block `{}` is sufficient.
         <properties>
             <property name="xpath">
                 <value>
-//DoStatement[Expression/PrimaryExpression/PrimaryPrefix/Literal/BooleanLiteral] |
+<![CDATA[
+//DoStatement[Expression/(.|*)/PrimaryExpression/PrimaryPrefix/Literal/BooleanLiteral] |
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/bestpractices.xml:2175]: The path `(.|*)` will match 0 or exactly one level. Maybe we can write it as `Expression//PrimaryExpression...`, which will match 0 or more levels. That however might create more false positives, not sure.

Could you also check, what effect parentheses have? E.g. `do {} while ( (false|false) )`

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for the PR!

* Please also add cases for fixing the `while` loops. The rule is about both do- and while-loops, so fixing only one part makes the rule assymetric. Eg. `while(false || false) {}` should be flagged as well.
* Looking that the regression tester reports (https://chunk.io/pmd/84de32cf73d8463c934e520fbe31e1c0/diff1/index.html) I find these examples:
    * https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexity.java#L145
        * I'm not sure the rule should actually flag this. The and condition `&& true` is redundant, however the first part of the condition is relevant. The rule would suggest to write the loop as `while(a<0) {}`. But that doesn't seem to fit the rule name "WhileLoopWithLiteralBoolean", and converting a do-loop to a while-loop might change the behaviour depending on the condition. Only when the whole condition can be reduced to a boolean at compile time, the rule should be triggered.
        * Yes, I believe this is a false positive
    * https://github.com/checkstyle/checkstyle/blob/checkstyle-9.1/src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexity.java#L150
        * This is a bit different here. The condition `|| true` make the first condition obsolete, because the whole condition will always evaluate to true, hence it's actually the `do {} while(true)` case.
        * This violation is correct now and you change actually fixed a false negative.
    * https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/sun/security/x509/AVA.java#L482
        * This is a false positive
        * We probably should restrict the operators to be only '|' or '||'
    * Please also add additional tests for other combinations, e.g. `do {} while (true || false)`, `do {} while (true && false)`, `do {} while (false || true)`, `do {} while (false && true)`. You can combine all the variants into a single test case and verify with `expected-problems` and `expected-linenumbers` (multiple lines can be provided as a comma separated list).

**Pull request merged**